torchaudio/prototype/rnnt_decoder.py

```diff
@@ -0,0 +1,361 @@
+        return new_hypos
+
+    def _search(
+        self, enc_out: torch.Tensor, hypo: Optional[Hypothesis], beam_width: int,
+    ) -> List[Hypothesis]:
+        n_time_steps = enc_out.shape[1]
+        device = enc_out.device
+
+        a_hypos: List[Hypothesis] = []
+        b_hypos = self._init_b_hypos(hypo, device)
+        for t in range(n_time_steps):
+            a_hypos = b_hypos
+            b_hypos = torch.jit.annotate(List[Hypothesis], [])
+            key_to_b_hypo: Dict[str, Hypothesis] = {}
+            symbols_current_t = 0
+
+            while a_hypos:
+                next_token_probs = self._gen_next_token_probs(
+                    enc_out[:, t: t + 1], a_hypos, device
+                )
+                next_token_probs = next_token_probs.cpu()
+                b_hypos = self._gen_b_hypos(
+                    b_hypos, a_hypos, next_token_probs, key_to_b_hypo,
+                )
+
+                if symbols_current_t == self.step_max_tokens:
+                    break
+
+                a_hypos = self._gen_a_hypos(
+                    a_hypos, b_hypos, next_token_probs, t, beam_width, device,
+                )
+                if a_hypos:
+                    symbols_current_t += 1
+
+            _, sorted_idx = torch.tensor(
+                [self.hypo_sort_key(hypo) for hypo in b_hypos]
+            ).topk(beam_width)
+            b_hypos = [b_hypos[idx] for idx in sorted_idx]
+
+        return b_hypos
+
+    def forward(
+        self, input: torch.Tensor, length: torch.Tensor, beam_width: int
+    ) -> List[Hypothesis]:
+        r"""Performs beam search for the given input sequence.
+
+        T: number of frames;
+        D: feature dimension of each frame.
+
+        Args:
+            input (torch.Tensor): sequence of input frames, with shape (1, T, D).
+            length (torch.Tensor): number of valid frames in input
+                sequence, with shape (1,).
+            beam_width (int): beam size to use during search.
+
+        Returns:
+            List[Hypothesis]: top-``beam_width`` hypotheses found by beam search.
+        """
+        assert (
+            len(input.shape) == 3 and input.shape[0] == 1
+        ), "input must be of shape (1, T, D)"
+        assert length.shape == (1,), "length must be of shape (1,)"
+        enc_out, _ = self.model.transcribe(input, length)
+        return self._search(enc_out, None, beam_width)
+
+    @torch.jit.export
+    def infer(
+        self,
+        input: torch.Tensor,
+        length: torch.Tensor,
+        beam_width: int,
+        state: Optional[List[List[torch.Tensor]]] = None,
+        hypothesis: Optional[Hypothesis] = None,
+    ) -> Tuple[List[Hypothesis], List[List[torch.Tensor]]]:
+        r"""Performs beam search for the given input sequence in streaming mode.
+
+        T: number of frames;
+        D: feature dimension of each frame.
+
+        Args:
+            input (torch.Tensor): sequence of input frames, with shape (1, T, D).
```

Comment:
Is the first dimension something naturally originated from the RNN-T model output?
Or is it there to ease the following hypothesis generation process?

If it's for the later, and if users have to do extra `.unsqueeze(0)` in the client code, then I think it's more straightforward to accept 2D input. otherwise it's fine.

Reply:
good point — this is more a consequence of the rnn-t transcription network expecting an input of shape (B, T, D), but we don't have to make users adhere to this. perhaps accept both (1, T, D) and (T, D)?

Reply:
sounds good to me.